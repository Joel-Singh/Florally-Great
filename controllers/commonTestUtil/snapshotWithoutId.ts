import { Document } from "mongoose";

export function snapshotWithoutId(document: Document, snapshotName: string) {
  document = document.toObject();
  delete document._id;
  expect(document).toMatchSnapshot(snapshotName);
}
