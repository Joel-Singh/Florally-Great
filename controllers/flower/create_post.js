const asyncHandler = require('express-async-handler')
const path = require('path')
const he = require('he')
const Flower = require(path.join(appRoot, 'models', 'flower.js'))
const Region = require(path.join(appRoot, 'models', 'region.js'))
const renderFlowerForm = require('./util/renderFlowerForm')

const { body, validationResult } = require("express-validator");

const validationFunctions = [
  body('name')
    .trim()
    .isLength({ min: 1}).withMessage(`Name can't be empty`)
    .escape(),

  body('description')
    .trim()
    .isLength({ min: 1}).withMessage(`Description can't be empty`)
    .escape(),

  body('numberInStock')
    .trim()
    .isLength({ min: 1}).withMessage(`Number in stock can't be empty`)
    .isNumeric().withMessage('Number in stock needs to be a number'),

  body('price')
    .trim()
    .isLength({ min: 1}).withMessage(`Price can't be empty`)
    .matches(/\$[0-9]+\.[0-9][0-9]/).withMessage('Price needs to be in $x.xx format, e.g $3.86 or $287.00'),
]

const validateInput = async (req, res, next) => {
  const errors = validationResult(req)
  if (!errors.isEmpty()) {
    await renderFlowerForm(
      res,
      next,
      { errors: errors.array()}
    )
    return
  }
  next();
}

const checkDuplicateFlower = asyncHandler(async (req, res, next) => {
  const { name } = req.body
  const flowerExists = await Flower.findOne({ name }).exec()
  if (flowerExists) {
    await renderFlowerForm(
      res,
      next,
      { errors: [{ msg: 'Flower already exists'}] }
    )
    return
  }
  next();
})

const saveFlower = asyncHandler(async (req, res, next) => {
    const { name, description, price, numberInStock, region } = req.body
    const priceToNumber = parseFloat(price.slice(1))

    const convertedName = he.decode(name)
    const convertedDescription = he.decode(description)

    const flower = new Flower(
      {
        name: convertedName,
        description: convertedDescription,
        price: priceToNumber,
        numberInStock,
        region
      }
    )
    await flower.save()
    res.redirect(flower.url)
  }
)

module.exports = [
  validationFunctions,
  validateInput,
  checkDuplicateFlower,
  saveFlower
]
