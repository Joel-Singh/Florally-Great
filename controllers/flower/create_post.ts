import validateFlowerRequest from "./util/validateFlowerRequest";
import renderFlowerForm from "./rendersWithDefaultLocals/renderFlowerForm";
import getFlowerModelDataFromReqBody from "./util/getFlowerModelDataFromReqBody";
import Flower, { IFlowerProperties } from "../../models/flower";
import getCreatePostMiddleware from "../template/create_post";
import {
  FlowerFormPrepopulatedValues,
  RequestWithFlowerFormData,
} from "../../views/flowers/flowerFormData";

async function saveFlower(flowerProperties: IFlowerProperties) {
  const flower = new Flower(flowerProperties);
  await flower.save();
  return flower;
}

function getPreviousDataFromReqBody(
  req: RequestWithFlowerFormData,
): FlowerFormPrepopulatedValues {
  const { name, description, price, regionID, numberInStock } = req.body;
  return {
    name,
    description,
    price,
    region: regionID.toString(),
    numberInStock,
  };
}

export default getCreatePostMiddleware(
  validateFlowerRequest,
  renderFlowerForm,
  getFlowerModelDataFromReqBody,
  getPreviousDataFromReqBody,
  saveFlower,
);
