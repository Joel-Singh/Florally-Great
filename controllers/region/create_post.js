const asyncHandler = require("express-async-handler");
const path = require("path");
const { default: Region } = require(path.join(appRoot, "models", "region.ts"));

const { body, validationResult } = require("express-validator");

module.exports = [
  body("name", `Name can't be empty`).trim().isLength({ min: 1 }).escape(),

  body("description", `Description can't be empty`)
    .trim()
    .isLength({ min: 1 })
    .escape(),

  asyncHandler(async (req, res, next) => {
    const errors = validationResult(req);

    if (!errors.isEmpty()) {
      res.render("regions/region_form", {
        errors: errors.array(),
      });
      return;
    } else {
      const { name, description } = req.body;
      const regionExists = await Region.findOne({ name }).exec();

      if (regionExists) {
        res.render("regions/region_form", {
          errors: [{ msg: "Region already exists" }],
        });
      } else {
        const region = new Region({ name, description });
        await region.save();
        res.redirect(region.url);
      }
    }
  }),
];
