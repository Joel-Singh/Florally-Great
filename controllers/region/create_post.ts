import asyncHandler from "express-async-handler";
import { Request, Response, NextFunction } from "express";
import { body, validationResult } from "express-validator";
import Region, {
  IRegionDocument,
  IRegionProperties,
} from "../../models/region";
import getCreatePostMiddleware from "../template/create_post";
import {
  RegionFormLocals,
  RequestWithRegionFormData,
} from "../../views/regions/regionFormInterfaces";

const validators = [
  body("name", `Name can't be empty`).trim().isLength({ min: 1 }).escape(),
  body("description", `Description can't be empty`)
    .trim()
    .isLength({ min: 1 })
    .escape(),
];

function renderForm(res: Response, locals: RegionFormLocals) {
  res.render("regions/region_form", {
    ...locals,
  });
}

function getModelDataFromReqBody(
  req: RequestWithRegionFormData
): IRegionProperties {
  const { name, description } = req.body;
  return { name, description };
}

function getPreviousDataFromReqBody(
  req: RequestWithRegionFormData
): IRegionProperties {
  const { name, description } = req.body;
  return { name, description };
}

async function saveDocument(properties: IRegionProperties) {
  const region: IRegionDocument = new Region(properties);
  await region.save();
  return region;
}

export default getCreatePostMiddleware(
  validators,
  renderForm,
  getModelDataFromReqBody,
  getPreviousDataFromReqBody,
  saveDocument
);
