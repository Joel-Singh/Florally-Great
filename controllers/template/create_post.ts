import { NextFunction, Request, Response } from "express";
import {
  ValidationChain,
  ValidationError,
  validationResult,
} from "express-validator";
import asyncHandler from "express-async-handler";

export default function getCreatePostMiddleware<
  DocumentProperties,
  Document extends { url: string },
>(
  validators: ValidationChain[],
  renderForm: (
    res: Response,
    locals: { errors: ValidationError[] },
  ) => Promise<void>,
  getModelDataFromReqBody: (res: Request) => DocumentProperties,
  saveDocument: (
    docProperties: DocumentProperties,
  ) => Promise<Document> | Promise<Error>,
) {
  return [
    validators,
    // @ts-ignore
    asyncHandler(async function (
      req: Request,
      res: Response,
      next: NextFunction,
    ) {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        await renderForm(res, { errors: errors.array() });
      } else {
        const document = await saveDocument(getModelDataFromReqBody(req));
        if ("url" in document) {
          res.redirect(document.url);
        } else {
          throw new Error(
            "Stinkers, did the document saved not have a url property or errored out?",
          );
        }
      }
    }),
  ];
}
