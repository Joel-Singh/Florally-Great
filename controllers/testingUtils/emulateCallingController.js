module.exports = async function (controller, reqProperties, resProperties) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);

  await controller(fakeReq, fakeRes);

  return {
    fakeReq,
    fakeRes,
    getRenderInformation,
  };

  function getFakeMiddleware(options) {
    const fakeReq = {
      ...reqProperties,
    };

    const fakeRes = {
      render: jest.fn(),
      redirect: jest.fn(),
      ...resProperties,
    };

    return { fakeReq, fakeRes };
  }
};

function getRenderInformation(fakeRes) {
  const renderCalls = fakeRes.render.mock.calls;
  if (renderCalls.length === 0) throw new Error("Render hasn't been called!");

  const renderCall = renderCalls[0];
  return {
    view: renderCall[0],
    locals: renderCall[1],
  };
}
