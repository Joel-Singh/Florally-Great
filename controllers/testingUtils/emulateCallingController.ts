import { Request, Response } from "express";

type nestedFunctionArray = Array<Function | nestedFunctionArray>;
export default async function emulateCallingController(
  controller: nestedFunctionArray | Function,
  reqProperties: Partial<Request> = {},
  resProperties: Partial<Response> = {},
) {
  const { fakeReq, fakeRes } = getFakeMiddleware(reqProperties, resProperties);
  const mockNext = jest.fn();

  if (!Array.isArray(controller)) {
    await controller(fakeReq, fakeRes, mockNext);
  } else {
    // @ts-ignore
    controller = controller.flat(Infinity) as Array<Function>;
    await runMiddlewareArray(
      controller as Array<Function>,
      fakeReq,
      fakeRes,
      mockNext,
    );
  }

  return {
    fakeReq,
    fakeRes,
    mockNext,
    getRenderInformation: createMockInfoGetter(
      fakeRes.render,
      "view",
      "locals",
    ),
    getRedirectInformation: createMockInfoGetter(
      fakeRes.redirect,
      "redirectPage",
    ),
    getMockNextInformation: createMockInfoGetter(mockNext, "errorOrEmpty"),
  };
}

function getFakeMiddleware(
  reqProperties: Partial<Request>,
  resProperties: Partial<Response>,
) {
  const fakeReq = {
    ...reqProperties,
  };

  const fakeRes = {
    ...resProperties,
    render: jest.fn(),
    redirect: jest.fn(),
    send: jest.fn(),
  };

  return { fakeReq, fakeRes };
}

function createMockInfoGetter<T extends Record<string, any>>(
  mockFunction: jest.Mock,
  ...namesInArgumentOrder: Array<keyof T>
) {
  return function () {
    const mockCalls = mockFunction.mock.calls;
    if (mockCalls.length === 0)
      throw new Error("Mock function hasn't been called!");

    const mockCall = mockCalls[0];
    const mockInfo: Partial<T> = {};

    namesInArgumentOrder.forEach((name, index) => {
      mockInfo[name] = mockCall[index];
    });

    return mockInfo as T;
  };
}

async function runMiddlewareArray(
  middlewares: Array<Function>,
  req: Partial<Request>,
  res: Partial<Response>,
  next: ReturnType<typeof jest.fn>,
) {
  for (const middleware of middlewares) {
    let previousCallLength = next.mock.calls.length;
    await middleware(req, res, next);
    let currentCallLength = next.mock.calls.length;
    let nextWasCalled = currentCallLength > previousCallLength;
    if (!nextWasCalled) break;
  }
}
