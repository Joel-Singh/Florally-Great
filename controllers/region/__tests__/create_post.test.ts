import Region from "../../../models/region";
import getValidRegionData from "../../../utils/dummyData/getValidData/getValidRegionData";
import { snapshotWithoutId } from "../../commonTestUtil/snapshotWithoutId";
import emulateCallingController from "../../testingUtils/emulateCallingController";
import create_post from "../create_post";

test("Redirects and saves region with valid data", async () => {
  const { getRedirectInformation } = await emulateCallingController(
    create_post,
    {
      body: getValidRegionData(),
    }
  );

  const savedRegion = (await Region.find({}))[0];
  snapshotWithoutId(savedRegion, "Region saved");
  expect(getRedirectInformation()).toMatchSnapshot("Correct Redirect");
});

test("Renders region form with errors with invalid data", async () => {
  const { getRenderInformation } = await emulateCallingController(create_post, {
    body: {
      ...getValidRegionData(),
      name: "",
    },
  });

  expect(await Region.find({})).toStrictEqual([]);

  const {
    view,
    locals: { errors },
  } = getRenderInformation();

  expect(errors.length).toBeGreaterThanOrEqual(1);
  expect(view).toMatchInlineSnapshot(`"regions/region_form"`);
});

test("When rendering with error, renders previous inputs", async () => {
  const { getRenderInformation } = await emulateCallingController(create_post, {
    body: {
      ...getValidRegionData(),
      name: "",
    },
  });

  expect(await Region.find({})).toStrictEqual([]);
  const {
    locals: { prepopulatedValues },
  } = getRenderInformation();

  expect(prepopulatedValues).toMatchInlineSnapshot(`
    {
      "description": "Description",
      "name": "",
    }
  `);
});
