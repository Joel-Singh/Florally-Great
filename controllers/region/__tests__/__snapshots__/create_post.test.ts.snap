// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redirects and saves region with valid data: Correct Redirect 1`] = `
{
  "redirectPage": "/regions/Name1",
}
`;

exports[`Redirects and saves region with valid data: Region saved 1`] = `
{
  "__v": 0,
  "description": "Description",
  "name": "Name1",
}
`;
